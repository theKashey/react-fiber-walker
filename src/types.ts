export type TagType = 'host' | 'fc' | 'class' | 'lazy' | 'memo' | 'forward' | 'provider' | 'consumer' | 'portal' | '';

export interface Stack {
  type: any;
  props: object;
  elementType: any;
  instance: any;
  fiber: any;

  children: Stack[];

  tag: TagType;
}
