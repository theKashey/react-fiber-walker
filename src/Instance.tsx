import * as React from 'react';

export class Instance extends React.Component<{
  children: React.ReactNode;
  assignInstance: (instance: any) => void;
}> {
  componentDidMount() {
    this.props.assignInstance(this);
  }

  componentWillMount() {
    this.props.assignInstance(null);
  }

  render() {
    return this.props.children as any;
  }
}