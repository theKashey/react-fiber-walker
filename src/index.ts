export {getFiberStack} from './getStack';
export {Instance} from './Instance';

export {walk} from './walker';