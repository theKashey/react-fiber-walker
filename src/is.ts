import * as React from "react";
import {TagType} from "./types";

const ContextType = React.createContext ? React.createContext(null) : null;
const ConsumerType = ContextType && ContextType.Consumer.$$typeof;
const ProviderType = ContextType && ContextType.Provider.$$typeof;
const MemoType = React.memo && React.memo(() => null).$$typeof;
const LazyType = React.lazy && React.lazy(() => null).$$typeof;
const ForwardType = React.forwardRef && React.forwardRef(() => null).$$typeof;

export const isContextConsumer = (type) =>
  type && typeof type === 'object' && '$$typeof' in type && type.$$typeof === ConsumerType && ConsumerType;
export const isContextProvider = (type) =>
  type && typeof type === 'object' && '$$typeof' in type && type.$$typeof === ProviderType && ProviderType;
export const isMemoType = (type) =>
  type && typeof type === 'object' && '$$typeof' in type && type.$$typeof === MemoType && MemoType;
export const isLazyType = (type) =>
  type && typeof type === 'object' && '$$typeof' in type && type.$$typeof === LazyType && LazyType;
export const isForwardType = (type) =>
  type && typeof type === 'object' && '$$typeof' in type && type.$$typeof === ForwardType && ForwardType;

export const getType = (type: any, instance: any): TagType => {
  if (typeof type === 'string') return 'host';

  if (typeof type === 'function' && !instance) return 'fc'
  if (typeof type === 'function' && instance) return 'class'

  if (!!isLazyType(type)) return 'lazy';
  if (!!isMemoType(type)) return 'memo';
  if (!!isContextConsumer(type)) return 'consumer';
  if (!!isContextProvider(type)) return 'provider';
  if (!!isForwardType(type)) return 'forward';

  return '';
}