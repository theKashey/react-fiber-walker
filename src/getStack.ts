import {Component} from "react";
import {Stack} from "./types";
import {getType} from "./is";

function pushStack(node: any, stack: Stack) {
  const instance = typeof node.type === 'function' ? node.stateNode : stack;

  stack.type = node.type;
  stack.props = node.memoizedProps;
  stack.elementType = node.elementType || node.type;

  stack.children = [];

  stack.tag = getType(stack.type, instance);

  Object.defineProperty(stack, 'fiber', {
    value: node,
    enumerable: false,
  });

  Object.defineProperty(stack, 'instance', {
    value: instance,
    enumerable: false,
  });
}

function hydrateFiberStack(node: any, stack: Stack) {
  pushStack(node, stack);
  if (node.child) {
    let {child} = node;
    do {
      const childStack: Stack = {} as any;
      hydrateFiberStack(child, childStack);
      stack.children.push(childStack);
      child = child.sibling;
    } while (child);
  }
}

export const getInternalInstance = instance =>
  instance._reactInternalFiber || // React 16
  instance._reactInternalInstance || // React 15
  null;

export function getFiberStack(instance: Component): Stack {
  const rootNode: any = getInternalInstance(instance);
  const stack: Stack = {} as any;
  if (rootNode) {
    hydrateFiberStack(rootNode, stack);
  }

  return stack;
}