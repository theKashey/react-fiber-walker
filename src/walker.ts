import {Component} from "react";
import {Stack} from "./types";
import {getFiberStack} from "./getStack";

export type Visitor = (element: Stack, instance: any) => void;

const walkTree = (tree: Stack, visitor: Visitor) => {
  visitor(tree, tree.instance);
  tree.children.forEach(child => walkTree(child, visitor));
};

export const walk = (entry: Component, visitor: Visitor) => {
  const tree = getFiberStack(entry);
  walkTree(tree, visitor);
};