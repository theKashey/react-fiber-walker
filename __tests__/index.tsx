import * as React from 'react';
import {mount} from 'enzyme';
import {getFiberStack} from "../src";

describe('Specs', () => {
  it('stack', () => {
    class Comp extends React.Component {
      render() {
        return "42"
      }
    }

    class Thought extends React.Component {
      render() {
        return this.props.children;
      }
    }

    const Test = () => <div><Comp/></div>;
    const wrapper = mount(<Thought p1={1}><div><Test p2={2}/></div></Thought>);

    expect(getFiberStack(wrapper.instance())).toMatchSnapshot();

  });
});
