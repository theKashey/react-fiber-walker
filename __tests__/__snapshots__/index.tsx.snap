// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Specs stack 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "elementType": null,
                      "props": "42",
                      "tag": "",
                      "type": null,
                    },
                  ],
                  "elementType": [Function],
                  "props": Object {},
                  "tag": "class",
                  "type": [Function],
                },
              ],
              "elementType": "div",
              "props": Object {
                "children": <Comp />,
              },
              "tag": "host",
              "type": "div",
            },
          ],
          "elementType": [Function],
          "props": Object {
            "p2": 2,
          },
          "tag": "fc",
          "type": [Function],
        },
      ],
      "elementType": "div",
      "props": Object {
        "children": <Test
          p2={2}
        />,
      },
      "tag": "host",
      "type": "div",
    },
  ],
  "elementType": [Function],
  "props": Object {
    "children": <div>
      <Test
        p2={2}
      />
    </div>,
    "p1": 1,
  },
  "tag": "class",
  "type": [Function],
}
`;
